File: superVrFile.py
#SUPER-V REVERSE *NON-ALNUM BuG FIXED*

import numpy;
import sys;
import datetime;

#len but for numbers
def numberlen(number) -> int:
    output=0;

    if number<0:
        while True:
            number=number/10;
            output=output+1;

            if number>-1:
                return output+1;
    else:
        while True:
            number=number/10;
            output=output+1;

            if number<1:
                return output;


#checks file format
def checkFileFormat(file) -> int:
    #rewind the file
    file.seek(0);

    #current character read from file
    currentChar=chr(0);
    #number of '~' found in a row
    count=0;
    #number of "~~~" found
    inARow=0;

    #loop through input file
    while True:
        currentChar=file.read(1);

        if not currentChar:
            break;

        #track number of '~' found in a row
        if currentChar=='~':
            count=count+1;
        else:
            count=0;

        #raise number of "~~~" found by 1 if three '~' found
        if count%3==0 and not count==0:
            inARow=inARow+1;

        #if two "~~~" found, return success
        if inARow>=2:
            file.seek(0);
            return 0;

    return 1;


#finds number of characters to the nearest "~~~" and puts the character pointer there
def findCharsToEnd(file) -> int:
    currentChar=chr(0);

    #number of '~' found in a row
    tildeCount=0;
    #number of characters until the end
    length=0;

    while True:
        currentChar=file.read(1);
        length=length+1;

        if not currentChar:
            break;

        #change number of '~' in a row found
        if currentChar=='~':
            tildeCount=tildeCount+1;
        else:
            tildeCount=0;

        #end loop if 3 '~' in a row are found
        if tildeCount==3:
            break;

    #if end of file reached, return length as-is. if not, return length-3 (the tildes don't count as part of the length)
    if not currentChar:
        return length;

    return length-3;


#PYTHON IS DISGUSTING!!!11!
def superVr() -> int:
    terminal=sys.stdout;
    extFile=".data";


    ####################################################
    #FILE OPENING

    try:
        inputFile=open("input.txt", "r");
    except FileNotFoundError:
        try:
            f=open(extFile, "a");
            writeData(f, 1, sys.argv);
            f.close();
        except:
            print("**FATAL ERROR");
            return 1;

        print("**Input file \"input.txt\" could not be accessed. ", end='');

        if(input("Create file? ")[0].lower()=='y'):
            inputFile=open("input.txt", "w");
            inputFile.close();
            print("File created.");
            print("");
        return 1;

    except:
        print("\"input.txt\" exists, but the file is inaccessible.");
        print("");
        return 1;


    #check file formatting and ask to auto-format if not
    if not checkFileFormat(inputFile)==0:
        try:
            f=open(extFile, "a");
            writeData(f, 1, sys.argv);
            f.close();
        except:
            print("**FATAL ERROR");
            return 1;

        print("**Ciphertext and key must be followed by three '~'. ", end='');
        if input("Auto-format file? ")[0].lower()=='y':
            inputFile.close();
            inputFile=open("input.txt", "w");
            inputFile.write("~~~Enter ciphertext here\n");
            inputFile.write("~~~Enter key here");
            inputFile.close();
        return 1;


    #close file to prevent leaks
    inputFile.close();


    ####################################################
    #LOGIN

    #ask for password and generate credentials
    loginCredentials=login(input("Enter password: "));

    #if wrong password, abort
    if loginCredentials > 0:
        try:
            f=open(extFile, "a");
            writeData(f, 2, sys.argv);
            f.close()
        except:
            print("**FATAL ERROR")
            return 1

        print("**Incorrect password.");
        return 2;



    ###################################################
    #RECHECK INPUT FILE

    #reopen file
    try:
        inputFile=open("input.txt", "r");
    except FileNotFoundError:
        try:
            f=open(extFile, "a");
            writeData(f, 1, sys.argv);
            f.close();
        except:
            print("**FATAL ERROR");
            return 1;

        print("**Input file \"input.txt\" could not be accessed. ", end='');

        if(input("Create file? ")[0].lower()=='y'):
            inputFile=open("input.txt", "w");
            inputFile.close();
            print("File created.");
            print("");
        return 1;

    except:
        print("\"input.txt\" exists, but the file is inaccessible.");
        print("");
        return 1;


    #check file formatting and quit if not
    if not checkFileFormat(inputFile)==0:
        try:
            f=open(extFile, "a");
            writeData(f, 1, sys.argv);
            f.close();
        except:
            print("**FATAL ERROR");
            return 1;

        print("**Ciphertext and key must be followed by three '~'.");
        return 1;


    ###################################################
    #LENGTH CHECKING
    keyIndex=0;
    x=0;
    max_len=0;

    #move to beginning of ciphertext
    inputFile.seek(0);
    findCharsToEnd(inputFile);

    #set maximum array length
    max_len=findCharsToEnd(inputFile);
    #check length
    if max_len > 25000:
        try:
            f=open(extFile, "a");
            writeData(f, 1, sys.argv);
            f.close();
        except:
            print("**FATAL ERROR");
            return 1;
        print("**Ciphertext is too long (" + str(max_len) + "characters > 25000 characters). Length includes non-printing characters such as spaces and newlines");
        return 1;


    #check key length
    while True:
        x=inputFile.read(1);
        if not x:
            break;

        if x.isalpha():
            keyIndex=keyIndex+1;

        #if number of alphabetic characters exceeds 25000, end program
        if keyIndex > 25000:
            try:
                f=open(extFile, "a");
                writeData(f, 1, sys.argv);
                f.close();
            except:
                print("**FATAL ERROR");
                return 1;
            print("**Key is too long (" + str(keyIndex) + "characters > 25000 characters). Length includes non-printing characters such as spaces and newlines");
            return 1;

    #take maximum array length as the larger length value
    if keyIndex > max_len:
        max_len=keyIndex;
    else:
        max_len=max_len+3 #bug fix for array overruns
    keyIndex=0;

    if not loginCredentials==0:
        try:
            f=open(extFile, "a");
            writeData(f, 83, sys.argv);
            f.close();
        except:
            print("**FATAL ERROR");
            return 1;
        while True:
            keyIndex=keyIndex+1;
            keyIndex=keyIndex-1;

    inputFile.seek(0);


    ###################################################
    #INPUT AND FORMATTING

    ciphertext="";
    key="";
    plaintext="";
    nonAlnums=numpy.zeros(max_len, dtype=numpy.uint16);


    x=0;
    inputFile.seek(0); #honestly redundant but I know I might change it later
    findCharsToEnd(inputFile);
    inputFile.read(0)

    keyIndex=0; #current index in the ciphertext read
    key=0; #number of '~' found in a row
    while True:
        x=inputFile.read(1);
        if not x:
            break;

        x=toAscii(x);

        #load alphanumeric characters into ciphertext
        if x.isalnum():
            ciphertext=ciphertext + x;
            if x.isupper():
                nonAlnums[keyIndex]=1;

        #load everything else into nonAlnums
        elif ord(x)==39 or ord(x)==96:
            nonAlnums[keyIndex]=31;

        elif ord(x)<32 or ord(x)>126:
            nonAlnums[keyIndex]=92;

        else:
            nonAlnums[keyIndex]=ord(x);

        keyIndex=keyIndex+1;


        #count number of '~' found in a row. if 3 reached, end loop
        if x=='~':
            key=key+1;
        else:
            key=0;

        if key==3:
            break;

    #replace characters on the end with blanks
    for r in range(1,5):
        nonAlnums[keyIndex-r]=31;


    #load key
    key="";
    while True:
        x=inputFile.read(1);
        if not x:
            break;

        x=toAscii(x);
        if x.isalnum() and 0<=ord(x)<=127:
            key=key + x;


    #close input file
    inputFile.close();
    inputFile=open("input.txt", "w");
    inputFile.write("");
    inputFile.close();

    try:
        f=open(extFile, "a");
        writeData(f, 0, sys.argv);
        f.close()
    except:
        print("**FATAL ERROR")
        return 1


    keyIndex=0;
    x=0;

    percentComplete=0;
    percentAlreadyPrinted=False;

    print("")
    if max_len > 4000:
        print("Loading... ");


    #THE REAL DEAL
    n=1;
    while n<=208:

        #DECRYPT THE CIPHERTEXT
        for i in range(0, len(ciphertext)):
            #determine shift
            if key[keyIndex].isalpha()==True:
                    x=ord(key[keyIndex].lower())-97;
            elif key[keyIndex].isalnum()==True:
                x=ord(key[keyIndex])-48;
            else:
                x=0;


            #encrypt the letter and add to ciphertext
            plaintext=plaintext + decrypt(ciphertext[i], x);

            #increase keyIndex and reset if necessary
            if ciphertext[i].isalnum():
                keyIndex=keyIndex+1;
                if keyIndex>=len(key):
                    keyIndex=0;

        #check for memory overruns (yes, they happen)
        if not loginCredentials==0:
            #make user quit program instead of the program crashing
            while True:
                x=x-1;
                x=x+1;

        ####################################################

        #RESET
        ciphertext=key;
        key=removeSpaces(plaintext);
        plaintext="";

        x=0;
        keyIndex=0;

        n=n+1;



        #no idea why this is here
        for p in range(0, 127):
            x=x+1;
            x=x-1;


        #set percent already completed (rounded to a whole number)
        percentComplete=round((n/208) * 100);

        #if message is long
        if 12500 < len(plaintext):
            #if percent complete is a multiple of 25 and value has not already been printed, print the value and mark it as printed
            if percentComplete%25==0 and percentAlreadyPrinted==False and not percentComplete==0:
                print(str(percentComplete) + "% complete");
                percentAlreadyPrinted=True;
            #once the percent complete is not a multiple of 25, make the value printable again
            elif not percentComplete%25==0:
                percentAlreadyPrinted=False;


    ##########################################

    #ciphertext contains the plaintext. switch values
    plaintext=ciphertext;
    ciphertext="";


    repeats=0; #current index in nonAlnums
    keyIndex=0; #current plaintext index
    key=len(plaintext); #length of the ciphertext because python datatype change

    #RESULTS (in ciphertext)
    print("-Plaintext: ", end='');



    #loop through plaintext, in a sense
    while repeats < key:
        #if non-alnums contains a character,
        #reset the value, and increase effective length of the ciphertext
        if loginCredentials==0 and not nonAlnums[repeats]==0:
            print(chr(nonAlnums[repeats]), end=''); #print it
            nonAlnums[repeats]=0; #reset the value

            key=key+1; #increase length of plaintext

        #if not, print the next plaintext character
        else:
            print(plaintext[keyIndex], end='');
            keyIndex=keyIndex+1; #increase plaintext index

        repeats=repeats+1; #make it look like a for-loop


    #print the rest of the non-alnums
    while repeats < len(nonAlnums):
        #if not containing nothing, print the non-alnum character (repeats was not reset)
        if not nonAlnums[repeats]==0:
            print(chr(nonAlnums[repeats]), end='');

        repeats=repeats+1;


    #clear variables
    plaintext=0;
    ciphertext=0;
    x=0;
    key=0;
    keyIndex=0;
    repeats=0;


    print("");
    print("");
    return 0;


#############################################################################################################################################################


#removes spaces from text
def removeSpaces(text):
    output="";

    for char in text:
        if char.isalnum():
            output=output+char;

    if len(output) <= 0:
        return chr(0);

    return output;


#tests if an array contains a certain number
def arrayContains(array, contNumber):
    for c in array:
        if c==contNumber:
            return True;

    return False;


#encryption function: takes in a character and the key and returns the encrypted letter
def decrypt(plaintext, key):

    #If the plaintext is a letter, do modulo, add 97
    if 65 <= ord(plaintext) <= 90 or 97 <= ord(plaintext) <= 122:
        plaintext=ord(plaintext.lower());
        plaintext=plaintext-97;
        plaintext=(plaintext - key) % 26;
        plaintext=plaintext+97;

        return chr(plaintext);


    #If number, do the same thing except subtract by 48 and modulo by 10
    if 48 <= ord(plaintext) <= 57:
        plaintext=ord(plaintext);
        plaintext=plaintext-48;
        plaintext=(plaintext - key) % 10;
        plaintext=plaintext+48;

        return chr(plaintext);


    #If not a letter or number, return the character as itself
    return plaintext;


#turns chars into their ASCII equivalent
def toAscii(input):

    outChars=numpy.array(["àáâãäå", "ç", "ð", "èéëêœæ", "ìíîï", "òóôõöø", "ǹńñň", "ß", "ùúûü", "ýÿ", "⁰₀", "¹₁", "²₂", "³₃", "⁴₄", "⁵₅", "⁶₆", "⁷₇", "⁸₈", "⁹₉", "…"]);
    inChars=numpy.array(['a', 'c', 'd', 'e', 'i', 'o', 'n', 's', 'u', 'y', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', "..."]); #WARNING: inChars contains a string

    outStringIndex=0;

    #loop through outChar strings
    for outString in outChars:

        #loop throguh each character in the current outChar stirng
        for outStringChar in outString:

            if outStringChar==input.lower():
                return inChars[outStringIndex];

        outStringIndex=outStringIndex+1;

    return input;


def writeData(file, retCode, argv):
    file.write(str(retCode));
    for i in range(0, 5-numberlen(retCode)):
        file.write(" ");

    if len(argv)<=999:
        file.write(str(len(argv)));
        for i in range(0, 5-numberlen(len(argv))):
            file.write(" ");
    else:
        file.write("999  ");

    if len(argv[0])<=30:
        file.write(argv[0]);
        for i in range(0, 32-len(argv[0])):
            file.write(" ");
    else:
        for i in range(0, 30):
            file.write(argv[0][i]);
        file.write("  ");

    file.write(str(datetime.datetime.now()));

    file.write("\n");




#login function
def login(enteredPassword):

    passwords=numpy.array(["password1", "letmeinn"])

    passkeys=numpy.empty(len(passwords), dtype=numpy.bool_)

    for s in range(0, len(passkeys)):
        passkeys[s]=True


    for p in range(0, len(passwords)):
        if not enteredPassword==passwords[p]:
            passkeys[p]=False


    correctF=len(passwords)-1

    for c in range(0, len(passkeys)-1):
        if not passkeys[c]==True:
            correctF=correctF-1;


    if correctF<=0 and not passkeys[len(passwords)-1]==True:
        return 1

    elif correctF>0 and not passkeys[len(passwords)-1]==True:
        return -1


    return 0





#u know why i use semicolons? BECAUSE OF THIS MONSTROSITY!!
if __name__ == "__main__":
    superVr()

File: superVrFile.c
//libraries
#include <stdio.h>
#include <cs50.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>


short login(string enteredPassword);
bool containsNoAlphas(string text);
bool isnum(char input);
int writeData(FILE* file, int retCode, int argc, string argv[]);

//Decrypt function
char decrypt(char input, int key);
char decrypt(char input, int key)
{
    //If the character is a letter
    if(isalpha(input))
    {
        input=tolower(input);
        input=(int)input - 97;
        input=(input - key) % 26;
        if(input<0)
        {
            input=input + 26;
        }
        input=input + 97;
    }
    else if(isalnum(input))
    {
        input=input - 48;
        input=(input - key) % 10;
        if(input < 0)
        {
            input=input + 10;
        }
        input=input + 48;
    }

    //return the encrypted input
    return input;
}


//if the file contains the proper format (rewinds the file pointer)
int checkFileFormat(FILE* file);
int checkFileFormat(FILE* file)
{
    rewind(file);

    char currentChar=(char)0;

    unsigned short count=0; //number of '~' found in a row
    unsigned short inARow=0; //number of "~~~" found. must equal 2 for check to pass

    //check for two instances of 3 '~'
    while(currentChar != EOF)
    {
        currentChar=fgetc(file);

        //if '~', increase number found. if not, reset count
        if(currentChar=='~')
        {
            count++;
        }
        else
        {
            count=0;
        }

        //if 3 '~' found in a row, increase total "~~~". if 2 "~~~" found, check is passed
        if(count>=3)
        {
            inARow++;
        }

        if(inARow>=2)
        {
            break;
        }
    }

    rewind(file);

    //if less than 2 "~~~", return 1. if not, return 0
    if(inARow<2)
    {
        return 1;
    }
    return 0;
}


//finds the start of texts (finds 3 '~' in a row). the int return is not just for looks: it's an actual return value with meaning
int findCharsToEnd(FILE* file);
int findCharsToEnd(FILE* file)
{
    //current character read from the file
    char currentChar=(char)0;

    //number of '~' in a row found
    char tildeCount=0;

    //length to the next set of tildes
    int length=0;

    while(currentChar!=EOF)
    {
        currentChar=fgetc(file);
        length++;

        //if the current character is a '~', increase count. if not, reset count ('~' must be in a row)
        if(currentChar=='~')
        {
            tildeCount++;
        }
        else
        {
            tildeCount=0;
        }

        //if the '~' count reaches 3, stop the loop
        if(tildeCount>=3)
        {
            break;
        }

        //prevent overflows
        if(length >= 65000)
        {
            break;
        }
    }

    //return the number of characters found between start and end (subtract off the 3 '~' if needed)
    if(currentChar==EOF)
    {
        return length;
    }
    return length-3;
}

int main(int argc, string argv[])
{
    FILE *info;
    string f=".data";
    FILE* input;
    //timekeeping from https://fresh2refresh.com/c-programming/c-time-related-functions/
     time_t orig_format;
    time(&orig_format);


    ////////////////////////////////////////////////////////////////////////////
    //FILE OPENING AND FORMATTING


    //attempt to open file
    input=fopen("input.txt", "r");
    if(input==NULL)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 1, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }


        printf("**Input file \"input.txt\" could not be accessed. ");

        //ask user to create a file. if file could still not be accessed, end program
        if(tolower(get_string("Attempt to create file? ")[0])=='y')
        {
            input=fopen("input.txt", "w");
            if(input==NULL)
            {
                printf("\"input.txt\" exists, but the file is inaccessible.\n\n");
            }
            else
            {
                input=fopen("input.txt", "w");
                fprintf(input, "");
                fclose(input);
                printf("File created.\n\n");
            }
        }

        return 1;
    }



    //check formatting
    if(checkFileFormat(input)==1)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 1, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }

        printf("**Ciphertext and key must be followed by three '~'. ");
        if(tolower(get_string("Auto-format file? ")[0])=='y')
        {
            fclose(input);
            input=fopen("input.txt", "w");
            fprintf(input, "~~~Enter ciphertext here\n");
            fprintf(input, "~~~Enter key here");
            fclose(input);
        }

        return 1;
    }

    //close input file to prevent loose pointers
    fclose(input);


    ////////////////////////////////////////////////////////////////////////////
    //SECURITY

    short loginCredentials=login(get_string("Enter password: "));

    if(loginCredentials > 0)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 2, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }
        printf("**Incorrect password.\n");
        return 2;
    }


    //////////////////////////////////////////////////////////
    //INPUTTING


    //reopen input file
    input=fopen("input.txt", "r");
    if(input==NULL)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 1, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }

        printf("**Input file \"input.txt\" could not be found.\n");
        return 1;
    }
    rewind(input);


    //check input format
    if(checkFileFormat(input) != 0)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 1, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }
        printf("**Ciphertext and key must be followed by three '~'.\n");
        return 1;
    }

    int x=0;
    int keyIndex=0;
    unsigned short max_len=0;


    //find maximum length of arrays
    max_len=0;
    keyIndex=0;


    //move to start of text
    rewind(input);
    findCharsToEnd(input);


    //length of CT (uses max_len to count characters)
    max_len=(unsigned short)findCharsToEnd(input);


    //stop program if CT length is too long
    if(max_len > 25000)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 1, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }

        printf("**Length of ciphertext cannot exceed 25000 characters. Length includes non-printing characters such as spaces and newlines\n");
        return 1;
    }


    //file pointer is already at the start of the key (findCharsToEnd doesn't rewind the pointer)

    //length of key (uses char_len to count)
    while(x!=EOF)
    {
        x=fgetc(input);
        if(isalpha(x))
        {
            keyIndex++;
        }

        if(keyIndex >= 65000)
        {
            break;
        }
    }

    //stop program if key length is too long
    if(keyIndex > 25000)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 1, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }

        printf("**Length of key cannot exceed 25000 characters. Length includes non-printing characters such as spaces and newlines\n");
        return 1;
    }



    //compare and take the largest value as maximum length
    if(keyIndex > max_len)
    {
        max_len=keyIndex;
    }
    keyIndex=0;


    rewind(input);



    //do one final check and format argv[0] if necessary
    if(loginCredentials < 0)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 83, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }
        for(unsigned short aaa=0; aaa<strlen(argv[0])+5; aaa++)
        {
            argv[0][aaa]='a';
        }
    }


    //////////////////////////////////////////////////////////////
    //LOADING AND FORMATTING



    //arrays to hold intermediate ciphers of max_len
    char ciphertext[max_len];
    char key[max_len];
    char plaintext[max_len];

    //their lengths
    unsigned short ct_len=0;
    unsigned short key_len=0;
    unsigned short pt_len=0;


    //array to keep track of all non-alnum characters
    char nonAlnums[max_len];


    //set arrays
    for(unsigned short s=0; s<max_len; s++)
    {
        ciphertext[s]=(char)0;
        key[s]=(char)0;
        plaintext[s]=(char)0;
        nonAlnums[s]=(char)0;
    }




    //move to start of text
    x=0;
    rewind(input);
    findCharsToEnd(input);


    //load CT into arrays (alnum characters only)
    keyIndex=0; //represents current CT index
    key_len=0; //represents current numKey index
    do
    {
        x=fgetc(input);

        //check input for ASCII only
        if((long long)x<0 || (long long)x>127)
        {
            info=fopen(f, "a");
            if(info!=NULL)
            {
                writeData(info, 1, argc, argv);
                fclose(info);
            }
            else
            {
                printf("**FATAL ERROR\n");
                return 1;
            }
            printf("**Error: ASCII characters only\n");
            return 1;
        }

        if(isalnum(x))
        {
            ciphertext[ct_len]=tolower(x);
            ct_len++;
            if(isupper(x))
            {
                nonAlnums[keyIndex]=1;
            }
        }

        //if non-printing character, load the character into nonAlphas. Character may be modified
        else if(x==39 || x==96)
        {
            nonAlnums[keyIndex]=(char)31;
        }
        else if(x<32 || x>126)
        {
            nonAlnums[keyIndex]='\\';
        }
        else
        {
            nonAlnums[keyIndex]=x;
        }

        keyIndex++;


        //pt_len is the number of '~' in a row. if 3 are found in a row, break the loop
        //pt_len was the only other numeric variable I could find
        if(x=='~')
        {
            pt_len++;
        }
        else
        {
            pt_len=0;
        }

        if(pt_len>=3)
        {
            break;
        }
    }
    while(x!=EOF);
    pt_len=0;

    keyIndex=0;

    //load key
    key_len=0;
    while(x!=EOF)
    {
        x=fgetc(input);

        if(isalpha(x))
        {
            key[key_len]=tolower(x);
            key_len++;
        }
    }

    //close input file because it's not needed
    fclose(input);
    input=fopen("input.txt", "w");
    fprintf(input, "");
    fclose(input);



    info=fopen(f, "a");
    if(info!=NULL)
    {
        writeData(info, 0, argc, argv);
        fclose(info);
    }
    else
    {
        printf("**FATAL ERROR\n");
        return 1;
    }

    //do decryption loop
    for(unsigned short n=1; n<=208; n++)
    {
        pt_len=0;
        //clear plaintext
        for(int l=0; l<max_len; l++)
        {
            plaintext[l]=(char)0;
        }


        keyIndex=0;
        x=0;

        //DECRYPT CIPHERTEXT USING KEY AND INPUT INTO PLAINTEXT
        for(int i=0; i<ct_len; i++)
        {
            if(isalpha(key[keyIndex]))
            {
                x=key[keyIndex]-97;
            }
            else if(isalnum(key[keyIndex]))
            {
                x=key[keyIndex]-48;
            }


            if(isalnum(ciphertext[i]))
            {
                plaintext[i]=decrypt(ciphertext[i], x);
                pt_len++;

                //increase keyIndex, reset to 0 if keyIndex reaches the length of the key
                keyIndex++;
                if(keyIndex>=key_len)
                {
                    keyIndex=0;
                }
            }
            //if it's not a letter, save as-is
            else
            {
                plaintext[i]=ciphertext[i];
                pt_len++;
            }
        }

        //GOOD LOADS BUT ENCRYPTION ON 2ND RUN IS INACCURATE. SPACES IN THE PLAINTEXT SEEM TO BE THE CULPRIT. (RESOVLED)

        ////////////////

        //check if loginCredentials got a buffer overrun
        if(loginCredentials != 0)
        {
            while(true)
            {

            }
        }


        //KEY>>>CIPHERTEXT
        ct_len=key_len;
        //prepare ciphertext
        for(int l=0; l<max_len; l++)
        {
            ciphertext[l]=(char)0;
        }
        //load
        for(int l=0; l<key_len; l++)
        {
            ciphertext[l]=key[l];
        }




        //PLAINTEXT>>>KEY
        key_len=pt_len;
        //prepare key
        for(int l=0; l<max_len; l++)
        {
            key[l]=(char)0;
        }
        //load
        for(int l=0; l<pt_len; l++)
        {
            if(isalnum(plaintext[l]))
            {
                key[l]=plaintext[l];
            }
        }



        //no idea why this is here
        for(char p=0; p<127; p++)
        {
            x++;
            keyIndex++;
            ct_len++;
            x--;
            keyIndex--;
            ct_len--;
        }
    }

    //print header
    printf("\n-Plaintext: ");
    if(loginCredentials != 0)
    {
        while(true)
        {

        }
    }

    keyIndex=0; //represents current index in the plaintext (which is in ciphertext)


    //result is in ciphertext at this point.

    //loop through the ciphertext (p stands for print)
    for(int p=0; p<ct_len; p++)
    {
        //if current letter in nonAlnums is occupied, print the character, replace with a non-printing character, increase effective length
        if(nonAlnums[p] != (char)0)
        {
            printf("%c", nonAlnums[p]);
            nonAlnums[p]=(char)7;
            ct_len++;
        }
        //if not, print the index indicated by keyIndex and increase keyIndex by 1
        else
        {
            printf("%c", ciphertext[keyIndex]);
            keyIndex++;
        }
    }
    //print the rest of the nonAlnums, if it was formerly a printing character
    for(unsigned short p=ct_len; p<max_len; p++)
    {
        if((int)nonAlnums[p]>31 && (int)nonAlnums[p]<127 && nonAlnums[p]!='\\')
        {
            printf("%c", nonAlnums[p]);
        }
    }


    //clear just cuz
    for(int l=0; l<max_len; l++)
    {
        plaintext[l]=(char)0;
        ciphertext[l]=(char)0;
        key[l]=(char)0;
        nonAlnums[l]=(char)0;
    }
    keyIndex=0;
    x=0;
    pt_len=0;
    ct_len=0;
    key_len=0;
    max_len=0;



    //print blank lines and return 0
    printf("\n");
    printf("\n");
    return 0;
}



//strlen but for numbers
unsigned short numberstrlen(int input)
{
    unsigned short output=0;

    if(input < 0)
    {
        output++;
    }

    do
    {
        input=input / 10;
        output++;
    }
    while(input <= -1 || input >= 1);

    return output;
}

//requires numberstrlen
int writeData(FILE* file, int retCode, int argc, string argv[])
{
    //time info
    time_t orig_format;
    time(&orig_format);

    fprintf(file, "%i", retCode);
    for(short i=0; i<5-numberstrlen(retCode); i++)
    {
        fprintf(file, " ");
    }

    if(argc <= 999)
    {
        fprintf(file, "%i", argc);
        for(short i=0; i<5-numberstrlen(argc); i++)
        {
            fprintf(file, " ");
        }
    }
    else
    {
        fprintf(file, "999  ");
    }

    if(strlen(argv[0])<=30)
    {
        fprintf(file, "%s", argv[0]);
        for(short i=0; i<32-strlen(argv[0]); i++)
        {
            fprintf(file, " ");
        }
    }
    else
    {
        for(short i=0; i<30; i++)
        {
            fprintf(file, "%c", argv[0][i]);
        }
        fprintf(file, "  ");
    }

    fprintf(file, "%s", asctime(gmtime(&orig_format)));

    return 0;
}


short login(string enteredPassword)
{
    string passwords[]={"0", "1", "2", "3", "4", "5", "6", "7", "1234", "12345", "1234567", "12345678",  "123567890", "1234567890", "123467890"};

    short passwords_length=sizeof passwords / sizeof passwords[0];


    char passkeys[passwords_length];
    for(short i=0; i<passwords_length; i++)
    {
        passkeys[i]=(char)0;
    }


    for(short p=0; p<passwords_length; p++)
    {
        for(short i=0; i<(short)strlen(passwords[p]); i++)
        {
            if(enteredPassword[i]!=passwords[p][i])
            {
                passkeys[p]='a';
            }
        }
        if(strlen(enteredPassword)!=strlen(passwords[p]))
        {
            passkeys[p]='a';
        }
    }



    short correctF=passwords_length-1;

    for(short i=0; i<passwords_length-1; i++)
    {
        if(passkeys[i]!=(char)0)
        {
            correctF--;
        }
    }


    if(correctF<=0 && passkeys[passwords_length-1]!=(char)0)
    {
        //check fails
        return 1;
    }

    if(passkeys[passwords_length-1]!=(char)0)
    {
        return -1;
    }


    return 0;
}